File: autotagger/autotagger/AutoTagger.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;

namespace autotagger
{
    /*
     * loan_amnt

purpose
car
credit_card
debt_consolidation
home_improvement
house
major_purchase
medical
moving
other
small_business
vacation
wedding

term
 36 months


emp_length
10+ years
< 1 year
10+ years
10+ years
1 year
3 years
8 years
9 years
4 years



home_ownership

RENT, OWN, MORTGAGE, OTHER.

annual_inc

loan_status

Current
Fully Paid
Charged Off
Default

     */


    public enum ExpressionType
    {
        Inclusive,
        Exclusive,
        Undefined
    }

    public class Expression
    {
        public int fieldIdx = -1;
        public string fieldName = "";
        public string value = null;
        public List<string> values = new List<string>();
        public ExpressionType type = ExpressionType.Undefined;
    }

    public class Rule
    {
        public string name = "";
        public List<Expression> expressions = new List<Expression>();
    }

    public class AutoTagger
    {
        private List<Rule> rules = new List<Rule>();

        private List<Expression> parseExpression(string s)
        {

            var t = s.Split('\t');

            if (t.Length < 3)
            {
                Console.WriteLine("Invalid expression for line '{0}'", s);
                Console.WriteLine("Expected format: field keyword value, [value2, value3, ...]");
                return null;
            }

            List<Expression> parsedExpressions = new List<Expression>();

            string fieldname = t[0];
            string keyword = t[1];

            if (keyword == "IS" || keyword == "EQUALS")
            {
                parsedExpressions.Add(new Expression
                {
                    fieldName = fieldname,
                    value = t[2],
                    type = ExpressionType.Inclusive
                });
            }
            else if (keyword == "IS ANY")
            {
                parsedExpressions.Add(new Expression
                {
                    fieldName = fieldname,
                    values = t[2].Split(',').ToList(),
                    type = ExpressionType.Inclusive
                });
            }
            else if (keyword == "ALL EXCEPT" || keyword == "IS NOT")
            {
                string[] tt = t[2].Split(',');
                for (int j = 0; j < tt.Length; j++)
                {
                    parsedExpressions.Add(new Expression
                    {
                        fieldName = fieldname,
                        value = tt[j],
                        type = ExpressionType.Exclusive
                    });
                }
            }
            else
            {
                Console.WriteLine("Invalid keyword for line '{0}'", s);
                Console.WriteLine();
            }

            return parsedExpressions;
        }

        private void readRules(string filename)
        {
            Console.WriteLine("Reading rules from {0}", filename);

            StreamReader sr = new StreamReader(filename);
            Rule def = null;
            while (!sr.EndOfStream)
            {
                string s = sr.ReadLine();
                if (string.IsNullOrWhiteSpace(s)) continue;

                if (s.Contains("\t") && def != null)
                {
                    List<Expression> e = parseExpression(s.Substring(1));
                    if (e != null) def.expressions.AddRange(e);
                }
                else
                {
                    if (def != null && def.expressions.Any()) rules.Add(def);
                    def = new Rule {name = s};
                }
            }

            if (def != null && def.expressions.Any()) rules.Add(def);
            sr.Close();

            Console.WriteLine("Rules read");
        }

        private void printRules()
        {
            Console.WriteLine("----------------------------------------");
            Console.WriteLine("\t\tRULES");
            Console.WriteLine("----------------------------------------");

            for (int i = 0; i < rules.Count; i++)
            {
                Rule rule = rules[i];

                Console.WriteLine("{0}. {1}", i + 1, rule.name);
                foreach (Expression e in rule.expressions)
                {
                    Console.WriteLine("\t{0}\t{1}\t{2}",
                        e.fieldName,
                        e.type,
                        (e.value ?? string.Join(",", e.values)));
                }
            }
            Console.WriteLine();
        }

        private void searchFieldIndexes(string[] headerRow)
        {
            Hashtable h = new Hashtable();
            for (int i = 0; i < headerRow.Length; i++) h[headerRow[i]] = i;

            for (int i = 0; i < rules.Count; i++)
            {
                Rule rule = rules[i];

                foreach (Expression e in rule.expressions)
                {
                    if (h[e.fieldName] != null) e.fieldIdx = (int) h[e.fieldName];
                }
            }
        }

        private bool matchesRule(string[] row, Rule rule)
        {
            foreach (Expression expression in rule.expressions)
            {
                if (expression.type == ExpressionType.Inclusive)
                {
                    if (expression.value != null && row[expression.fieldIdx] != expression.value) return false;
                    if (expression.values.All(value => row[expression.fieldIdx] != value)) return false;
                }
                else if(expression.type == ExpressionType.Exclusive)
                {
                    if (expression.value != null && row[expression.fieldIdx] == expression.value) return false;
                    if (expression.values.Any(value => row[expression.fieldIdx] == value)) return false;
                }
            }

            return true;
        }

        private void printResults(List<int> results)
        {
            Console.WriteLine("----------------------------------------");
            Console.WriteLine("\t\tRESULTS");
            Console.WriteLine("----------------------------------------");

            for (int i = 0; i < rules.Count; i++)
            {
                Rule rule = rules[i];

                Console.WriteLine("{0}. {1}\t\t{2}", i + 1, rule.name, results[i]);
            }

        }

        private void parseFile(string filename)
        {
            Console.WriteLine("Reading data from {0}", filename);

            int lines = 0;

            StreamReader sr = new StreamReader(filename);

            if (!sr.EndOfStream)
            {
                string s = sr.ReadLine();
                if (string.IsNullOrWhiteSpace(s))
                {
                    Console.WriteLine("Failed to receive header row");
                    return;
                }
                searchFieldIndexes(s.Split(','));
            }

            List<int> results = rules.Select(rule => 0).ToList();

            while (!sr.EndOfStream)
            {
                string s = sr.ReadLine();
                if (string.IsNullOrWhiteSpace(s)) continue;

                string[] row = s.Split(',');

                for (int i = 0; i < rules.Count; i++)
                {
                    if (matchesRule(row, rules[i])) results[i]++;
                }

                if (lines%100000 == 0) Console.Write("Processed {0} records\r", lines);
                lines++;
            }
            sr.Close();

            Console.WriteLine("Processed {0} records", lines);

            printResults(results);
        }

        public void run()
        {
            DateTime startTime = DateTime.Now;
            Console.WriteLine("Tagger started");

            readRules("rules.txt");
            printRules();

            parseFile("loan.csv");

            TimeSpan elapsed = DateTime.Now.Subtract(startTime);
            Console.WriteLine("Time elapsed: {0}", elapsed);
        }
    }
}

File: python/autotagger.py
"""This application does simplified categorization of CSV rows based on defined rules."""
#!/usr/bin/env python

import csv
from collections import namedtuple
import time
import sys

#class Point(namedtuple('Point', 'x y')):
#     __slots__ = ()
#     @property
#     def hypot(self):
#         return (self.x ** 2 + self.y ** 2) ** 0.5
#    def __str__(self):
#        return 'Point: x=%6.3f  y=%6.3f  hypot=%6.3f' % (self.x, self.y, self.hypot)
#

class Expression:
    """Expression"""
    field = ''
    fieldIdx = -1
    value = None
    inclusive = False

Rule = namedtuple("Rule", "name expressions")

def parse_expression(expression):
    """Parses expression from string"""

    expr_parts = expression.lstrip().split("\t")

    if len(expr_parts) < 3:
        print("Invalid expression for line '{0}'".format(expression))
        print("Expected format: field keyword value, [value2, value3, ...]")

        return None

    keyword = expr_parts[1]

    expr = Expression()
    expr.field = expr_parts[0]
    #expr.value = "," + expr_parts[2] + ","
    expr.value = expr_parts[2].split(",")

    if keyword == "IS" or keyword == "EQUALS" or keyword == "IS ANY":
        expr.inclusive = True

    elif keyword == "ALL EXCEPT" or keyword == "IS NOT":
        expr.inclusive = False

    else:
        print("Invalid keyword for line '{0}'".format(expression))
        print()

        return None

    return expr

def read_rules(filename):
    """Reads rules from file"""

    with open(filename, "r", encoding="utf-8-sig") as file_pointer:

        print("Reading rules from " + filename)

        rules = []
        rule = None

        for line in file_pointer:
            line = line.rstrip()

            if not line:
                continue

            if line.startswith("\t"):
                if not rule:
                    print("Found expression without rule, skipping. Expression: " + line)
                    continue

                expr = parse_expression(line)

                if expr:
                    rule.expressions.append(expr)

            else:
                rule = Rule(line, [])
                rules.append(rule)

            #print(line)

        file_pointer.close()

    return [r for r in rules if r.expressions]

def check_rule(entry, rule):
    """Checks entry for rule"""

    for expr in rule.expressions:

        #match = "," + entry[expr.fieldIdx] + "," in expr.value
        match = entry[expr.fieldIdx] in expr.value

        if (not match and expr.inclusive) or (match and not expr.inclusive):
            return False

    return True

def search_field_indexes(header_row, rules):
    """Applies fields index for faster lookup"""

    for idx, val in enumerate(header_row):
        for rule in rules:
            for expr in rule.expressions:
                if expr.field == val:
                    expr.fieldIdx = idx

    for rule in rules:
        for expr in rule.expressions:
            if expr.fieldIdx == -1:
                print("Rule {0} failed to find index for {1}".format(rule.name, expr.field))

    return

def parse_file(filename, rules):
    """Parses incoming file and tags the rows according rules"""

    print("Reading data from " + filename)

    results = {}
    for rule in rules:
        results[rule.name] = 0

    with open(filename, "r", encoding="utf-8-sig") as file_pointer:

        header_row = file_pointer.readline()

        if not header_row:
            print("Failed to receive header row")
            return

        search_field_indexes(header_row.split(","), rules)

        lines = 0

        for row in file_pointer:
            entry = row.split(",")

            for rule in rules:
                if check_rule(entry, rule):
                    results[rule.name] += 1

            if lines % 100000 == 0:
                print("Processed {0} records".format(lines), end="\r")
            lines += 1

    print("Processed {0} records".format(lines))

    return results

def print_results(rules, results):
    """Prints results on screen"""

    print("----------------------------------------")
    print("\t\tRESULTS")
    print("----------------------------------------")

    i = 0
    for rule in rules:
        i += 1
        print("{0: >3}. {1: <48} {2: >8}".format(i, rule.name, results[rule.name]))

    return

def print_rules(rules):
    """Prints rules on screen"""

    print("----------------------------------------")
    print("\t\tRULES")
    print("----------------------------------------")

    for rule in rules:
        print(rule.name)
        for expr in rule.expressions:
            mode = "INCLUSIVE" if expr.inclusive else "EXCLUSIVE"
            print("\t{0: <16}\t{1: <10}\t{2}".format(expr.field, mode, expr.value))
        print()

    print("{0} rules".format(len(rules)))

    return

def main(argv):
    """Main application block"""

    start_time = time.process_time()
    print("Tagger started")

    rules = read_rules("rules.txt")
    print_rules(rules)

    results = parse_file("loan.csv", rules)
    print_results(rules, results)

    elapsed = time.process_time() - start_time
    print("Time elapsed: {0}".format(elapsed))

if __name__ == "__main__":
    main(sys.argv)

# dumb execution string = 24.04